src/main/java/org/opensearch/security/ssl/http/netty/Netty4HttpRequestHeaderVerifier.java

```diff
@@ -0,0 +1,139 @@
+        this.restFilter = restFilter;
+        this.threadPool = threadPool;
+
+        this.injectUserEnabled = settings.getAsBoolean(ConfigConstants.SECURITY_UNSUPPORTED_INJECT_USER_ENABLED, false);
+        boolean disabled = settings.getAsBoolean(ConfigConstants.SECURITY_DISABLED, false);
+        if (disabled) {
+            sslConfig = new SSLConfig(false, false);
+        } else {
+            sslConfig = new SSLConfig(settings);
+        }
+        boolean client = !"node".equals(settings.get(OpenSearchSecuritySSLPlugin.CLIENT_TYPE));
+        this.passthrough = client || disabled || sslConfig.isSslOnlyMode();
+    }
+
+    @Override
+    public void channelRead0(ChannelHandlerContext ctx, DefaultHttpRequest msg) throws Exception {
+        // DefaultHttpRequest should always be first and contain headers
+        ReferenceCountUtil.retain(msg);
+
+        if (passthrough) {
+            ctx.fireChannelRead(msg);
+            return;
+        }
+
+        // Start by setting this value to false, only requests that meet all the criteria will be decompressed
+        ctx.channel().attr(SHOULD_DECOMPRESS).set(Boolean.FALSE);
+        ctx.channel().attr(IS_AUTHENTICATED).set(Boolean.FALSE);
+
+        final Netty4HttpChannel httpChannel = ctx.channel().attr(Netty4HttpServerTransport.HTTP_CHANNEL_KEY).get();
+        String rawPath = SecurityRestUtils.path(msg.uri());
+        String path = RestUtils.decodeComponent(rawPath);
+        Matcher matcher = PATTERN_PATH_PREFIX.matcher(path);
+        final String suffix = matcher.matches() ? matcher.group(2) : null;
+        if (API_AUTHTOKEN_SUFFIX.equals(suffix)) {
+            ctx.fireChannelRead(msg);
+            return;
+        }
+
+        final SecurityRequestChannel requestChannel = SecurityRequestFactory.from(msg, httpChannel);
+        ThreadContext threadContext = threadPool.getThreadContext();
+        try (ThreadContext.StoredContext ignore = threadPool.getThreadContext().stashContext()) {
+            injectUser(msg, threadContext);
+
+            boolean shouldSkipAuthentication = HttpMethod.OPTIONS.equals(msg.method())
+                || HEALTH_SUFFIX.equals(suffix)
+                || WHO_AM_I_SUFFIX.equals(suffix);
+
+            if (!shouldSkipAuthentication) {
+                // If request channel is completed and a response is sent, then there was a failure during authentication
+                restFilter.checkAndAuthenticateRequest(requestChannel);
+            }
+
+            ThreadContext.StoredContext contextToRestore = threadPool.getThreadContext().newStoredContext(false);
+            ctx.channel().attr(CONTEXT_TO_RESTORE).set(contextToRestore);
+
+            requestChannel.getQueuedResponse().ifPresent(response -> ctx.channel().attr(EARLY_RESPONSE).set(response));
+
+            boolean shouldDecompress = !shouldSkipAuthentication && requestChannel.getQueuedResponse().isEmpty();
```

Comment:
will this be authentication successful response or a generic auth response?

Reply:
Good question it a little misleading - 'response' are used for an early response such as an error response otherwise the security plugin forwards the request to be handled by the associated RestHandler - which would return the normal response that could include success/failures/etc..